pkg/controller/master/storagenetwork/storage_network.go

```
@@ -0,0 +1,851 @@
+			return nil
+		}
+
+		// retry again
+		logrus.Errorf("check net-attach-def existing error %v", err)
+		return err
+	}
+
+	if err := h.networkAttachmentDefinitions.Delete(namespace, name, &metav1.DeleteOptions{}); err != nil {
+		logrus.Errorf("remove nad error %v", err)
+		return err
+	}
+
+	setting.Annotations[OldNadStorageNetworkAnnotation] = ""
+	return nil
+}
+
+// earlyReturn bool
+// err, reconcile
+func (h *Handler) checkLonghornSetting(setting *harvesterv1.Setting) (bool, error) {
+	value, err := h.getLonghornStorageNetwork()
+	if err != nil {
+		logrus.Errorf("get Longhorn settings error: %v", err)
+		return true, err
+	}
+
+	currentNad := setting.Annotations[NadStorageNetworkAnnotation]
+	if currentNad == value {
+		// check if we need to restart monitoring pods
+		if err := h.checkPodStatusAndStart(); err != nil {
+			if _, updateConditionErr := h.setConfiguredCondition(setting, false, ReasonInProgress, MsgRestartPod); updateConditionErr != nil {
+				logrus.Errorf("update status error %v", updateConditionErr)
+				return true, updateConditionErr
+			}
+			return true, err
+		}
+
+		settingCopy := setting.DeepCopy()
+		if err := h.removeOldNad(settingCopy); err != nil {
+			logrus.Errorf("remove old nad error %v", err)
+			return true, err
+		}
+
+		// finish
+		if _, err := h.setConfiguredCondition(settingCopy, true, ReasonCompleted, ""); err != nil {
+			logrus.Errorf("update status error %v", err)
+			return true, err
+		}
+		// early return but no error, finish
+		return true, nil
+	}
+	return false, nil
+}
+
+// true: all detach
+func (h *Handler) checkLonghornVolumeDetached() (bool, error) {
+	volumes, err := h.longhornVolumeCache.List(util.LonghornSystemNamespaceName, labels.Everything())
+	if err != nil {
+		logrus.Errorf("volume error %v", err)
+		return false, err
+	}
+
+	for _, volume := range volumes {
+		if volume.Status.State != "detached" {
+			logrus.Infof("volume state: %v", volume.Status.State)
+			return false, nil
+		}
+	}
```

prefer to create the attached volume name list and store it in the conditions so UI can show it properly, and prefer to return the error at the end for retry.

---

I will log this first.